text/0000-return-position-impl-trait-in-traits.md

```diff
@@ -0,0 +1,285 @@
+
+[explainer]: https://rust-lang.github.io/impl-trait-initiative/explainer.html
+[apit]: https://rust-lang.github.io/impl-trait-initiative/explainer/apit.html
+[rpit]: https://rust-lang.github.io/impl-trait-initiative/explainer/rpit.html
+[rpit_trait]: https://rust-lang.github.io/impl-trait-initiative/explainer/rpit_trait.html
+
+## Example use case
+
+The use case for `-> impl Trait` in trait fns is similar to its use in other contexts: traits often wish to return "some type" without specifying the exact type. As a simple example that we will use through the RFC, consider the `NewIntoIterator` trait, which is a variant of the existing `IntoIterator` that uses `impl Iterator` as the return type:
+
+```rust
+trait NewIntoIterator {
+    type Item;
+    fn into_iter(self) -> impl Iterator<Item = Self::Item>;
+}
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+*This section assumes familiarity with the [basic semantics of impl trait in return position][rpit].*
+
+When you use `impl Trait` as the return type for a function within a trait definition or trait impl, the intent is the same: impls that implement this trait return "some type that implements `Trait`", and users of the trait can only rely on that. However, the desugaring to achieve that effect looks somewhat different than other cases of impl trait in return position. This is because we cannot desugar to a type alias in the surrounding module; we need to desugar to an associated type (effectively, a type alias in the trait).
+
+Consider the following trait:
+
+```rust
+trait IntoIntIterator {
+    fn into_int_iter(self) -> impl Iterator<Item = u32>;
+}
+```
+
+The semantics of this are analogous to introducing a new associated type within the surrounding trait;
+
+```rust
+trait IntoIntIterator { // desugared
+    type IntoIntIter: Iterator<Item = u32>;
+    fn into_int_iter(self) -> Self::IntoIntIter;
+}
+```
+
+(In general, this associated type may be generic; it would contain whatever generic parameters are captured per the generic capture rules given previously.)
+
+This associated type is introduced by the compiler and cannot be named by users.
+
+The impl for a trait like `IntoIntIterator` must also use `impl Trait` in return position:
+
+```rust
+impl IntoIntIterator for Vec<u32> {
+    fn into_int_iter(self) -> impl Iterator<Item = u32> {
+        self.into_iter()
+    }
+}
+```
+
+This is equivalent to specify the value of the associated type as an `impl Trait`:
```

Comment:
```suggestion
This is equivalent to specify the return type as an associated type, whose value is `impl Trait`:
```

I wouldn't write `the` in this phrase, I may be wrong but it kind of implies to me that we are talking about the associated type defined by the trait, which we've just said that it's unnameable.